samples/api/hello_triangle/hello_triangle.cpp

```diff
@@ -277,40 +277,42 @@ void HelloTriangle::init_device(Context &                        context,
 	std::vector<VkPhysicalDevice> gpus(gpu_count);
 	VK_CHECK(vkEnumeratePhysicalDevices(context.instance, &gpu_count, gpus.data()));
 
-	// Select the first GPU we find in the system.
-	context.gpu = gpus.front();
+	for(size_t i = 0;i < gpu_count && (context.graphics_queue_index < 0);i++) {
 
-	uint32_t queue_family_count;
-	vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, nullptr);
+		context.gpu = gpus[i];
 
-	if (queue_family_count < 1)
-	{
-		throw std::runtime_error("No queue family found.");
-	}
+		uint32_t queue_family_count;
+		vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, nullptr);
 
-	std::vector<VkQueueFamilyProperties> queue_family_properties(queue_family_count);
-	vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, queue_family_properties.data());
+		if (queue_family_count < 1)
+		{
+			throw std::runtime_error("No queue family found.");
+		}
 
-	uint32_t device_extension_count;
-	VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, nullptr));
-	std::vector<VkExtensionProperties> device_extensions(device_extension_count);
-	VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, device_extensions.data()));
+		std::vector<VkQueueFamilyProperties> queue_family_properties(queue_family_count);
+		vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, queue_family_properties.data());
 
-	if (!validate_extensions(required_device_extensions, device_extensions))
-	{
-		throw std::runtime_error("Required device extensions are missing, will try without.");
-	}
+		uint32_t device_extension_count;
+		VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, nullptr));
+		std::vector<VkExtensionProperties> device_extensions(device_extension_count);
+		VK_CHECK(vkEnumerateDeviceExtensionProperties(context.gpu, nullptr, &device_extension_count, device_extensions.data()));
 
-	for (uint32_t i = 0; i < queue_family_count; i++)
-	{
-		VkBool32 supports_present;
-		vkGetPhysicalDeviceSurfaceSupportKHR(context.gpu, i, context.surface, &supports_present);
+		if (!validate_extensions(required_device_extensions, device_extensions))
+		{
+			throw std::runtime_error("Required device extensions are missing, will try without.");
+		}
 
-		// Find a queue family which supports graphics and presentation.
-		if ((queue_family_properties[i].queueFlags & VK_QUEUE_GRAPHICS_BIT) && supports_present)
+		for (uint32_t i = 0; i < queue_family_count; i++)
 		{
-			context.graphics_queue_index = i;
-			break;
+			VkBool32 supports_present;
+			vkGetPhysicalDeviceSurfaceSupportKHR(context.gpu, i, context.surface, &supports_present);
+
+			// Find a queue family which supports graphics and presentation.
+			if ((queue_family_properties[i].queueFlags & VK_QUEUE_GRAPHICS_BIT) && supports_present)
```

Comment:
This is certainly a good thing to check.  It might make sense to include checking for which device is the "best" out of candidate devices or show how to create more selection criteria.  One big area that likely would be ideal to check is ensure the device is a discrete vs in built if there's more than one as it's popular for laptops to have both.

Reply:
Yeah. I have heard that some laptops even require both gpus be put in play, the discrete does the renders and then the app has to copy the completed framebuffer to the cpu-embedded gpu for presentation. So that affects not only which gpu, but how many and the creation options for the framebuffers, and the existence of a staging swapchain as a copy_dst.

I also agree about needing better logic to find the "best" gpu. But the definition of best needs some discussion. I have been playing with the idea of a scoring system, one element of which would be based on the types of available queues (bonus points to any gpu with a queue family that supports render, compute, and present). As a user though, my definition of best would more likely depend on either the more powerful gpu, or the one that happens to be directly connected to the display that contains the target frame; and I am not sure how to detect either case. Thoughts?